File: day023/delete.c
// https://leetcode.com/problems/delete-leaves-with-a-given-value/

#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};

struct TreeNode *removeLeafNodes(struct TreeNode *root, int target) {
  if (root) {
    root->left = removeLeafNodes(root->left, target);
    root->right = removeLeafNodes(root->right, target);
    if (!root->left && !root->right && root->val == target) {
      // free(root);
      root = NULL;
    }
  }
  return root;
}

void print_tree(struct TreeNode *root) {

  printf("Tree { val: %d, left: ", root->val);

  if (root->left)
    print_tree(root->left);
  else
    printf("None");

  printf(", right: ");

  if (root->left)
    print_tree(root->left);
  else
    printf("None");

  printf("}");
}

struct TreeNode *new_tree(int val, struct TreeNode *left,
                          struct TreeNode *right) {
  struct TreeNode *node = malloc(sizeof(struct TreeNode));
  *node = (struct TreeNode){val, left, right};
  return node;
}

int main(int argc, char *argv[]) {
  struct TreeNode *tree =
      new_tree(1, new_tree(2, new_tree(2, NULL, NULL), NULL),
               new_tree(3, new_tree(2, NULL, NULL), new_tree(4, NULL, NULL)));
  print_tree(tree);
  printf("\n");
  print_tree(removeLeafNodes(tree, 2));

  return 0;
}

File: 2024/day023/delete.rs
// https://leetcode.com/problems/delete-leaves-with-a-given-value/

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

impl From<(i32, Option<TreeNode>, Option<TreeNode>)> for TreeNode {
    fn from(value: (i32, Option<TreeNode>, Option<TreeNode>)) -> Self {
        Self {
            val: value.0,
            left: if value.1.is_some() {
                Some(Rc::new(RefCell::new(value.1.unwrap())))
            } else {
                None
            },
            right: if value.2.is_some() {
                Some(Rc::new(RefCell::new(value.2.unwrap())))
            } else {
                None
            },
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

struct Solution;
impl Solution {
    pub fn remove_leaf_nodes(
        root: Option<Rc<RefCell<TreeNode>>>,
        target: i32,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if let Some(ref cell) = root {
            let mut node = cell.borrow_mut();
            node.left = Self::remove_leaf_nodes(node.left.clone(), target);
            node.right = Self::remove_leaf_nodes(node.right.clone(), target);
            if node.left.is_none() && node.right.is_none() && node.val == target {
                return None;
            }
        }
        root
    }
}

fn main() {
    let tree = TreeNode::from((
        1,
        Some(TreeNode::from((2, Some(TreeNode::new(2)), None))),
        Some(TreeNode::from((
            3,
            Some(TreeNode::new(2)),
            Some(TreeNode::new(4)),
        ))),
    ));
    println!("{:#?}", tree);
    println!(
        "{:#?}",
        Solution::remove_leaf_nodes(Some(Rc::new(RefCell::new(tree))), 2)
    );
}
